### ISSUE TITLE
Implementation of project.parent

### ISSUE THREAD

**Issue by sonalkr132** [glittergallery/GlitterGallery#229]
It started with: "ActionView::CompiledTemplates::FORK" error on issue/show page
```Ruby
- if @project.id == @project.parent
          %span.badge.author PARENT
        - else FORK
```
As you can tell, it's and easy fix:
```Ruby
       - else
         %span.badge.author FORK
```
But There is lot more to it than that. Project create has used:
```Ruby
def create
    project = Project.new project_params
    project.user_id = current_user.id
    project.parent = project.id
```
project.id doesn't get created until after object is saved in database, so all the project.parents are just nil. I need suggestion solving this.
Now we can write that line
```Ruby
    project.parent = project.id
```
after 
```Ruby
    project.save
```
I don't that that would be very good idea. We would already be going against rails convention and in fork method we are setting child.parent = Project.parent. So every fork will end up having same parent.

Is thats what you guys want? Doesn't git work like  if I fork from a original_repo and if someone forks my repo then their parent should be my repo not the original repo. Confusing?
What I am suggesting is that parent relationship should be like:
original_repo <- my_repo <- my_repo_fork

what we can do is in create set project.id = nil
and in fork we use the id of project from which we are forking and not the parent.id of project being forked.

Let me know what you guys had on your mind and I will send you a PR.

**Comment by rohitpaulk** [comment]
Aah, thanks for the analysis. I agree with you, let's set project.parent to `nil` for a non-fork, and to the id of the project being forked (not the parent, the id of the object itself) for forks.

**Comment by sonalkr132** [comment]
Okay. should I migrate all the way down and make changes to "20131030132834_add_parent_to_projects" migration or should I add a new migration? I think you meant former?

**Comment by rohitpaulk** [comment]
Ah I'm sorry, I didn't think the change through - the migration should be a simple rename of a column, and hence won't change existing data. What I meant was - when you add the new migration, it's okay if it doesn't honor the values on previous projects. 

TLDR - add a new migration that renames the column.

**Comment by EmilyDirsh** [comment]
Actually, if you're forking a project from a fork, the parent of the new project _should_ be the parent id not the project id. That's not very clear. :P Take this diagram:
```
 ___________________________________________________
|   Project A [not a fork] id: 1, parent id: nil    |
 ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾|‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
 _________________________V_________________________
| Project B [fork of project A] id: 2, parent id: 1 |
 ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾|‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
 _________________________V_________________________
| Project C [fork of project B] id: 3, parent id: 1 |
 ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
```
For Project C, the correct parent id is `1`, the id of Project A, even though C is being forked from Project B. We want all subsequent forks to point back to the original project, even if it is being forked from a fork. I believe GitHub works this way as well.

**Comment by sonalkr132** [comment]
Are you sure thats how you want to keep things? Git doesn't work that way. I think we should discuss this further as setting parent to fork makes more sense than setting it to original repo. If I fork some's work then my repo would reflect the status(files and changes) of my parent fork and not the original repo. If i wanted my parent to be original repo then I would have forked original repo. @rohitpaulk seems to agree on this too.
But we can make changes to way it was, if that's what you want.
Except that we were setting project.parent_id = project.id, when project.id isn't created until after saving of record. I am sure we will think something, We should clear out what kind of relationship do we want between forks and repos first.

**Comment by rohitpaulk** [comment]
> I believe GitHub works this way as well.

That's not how GitHub works, here's an example. https://github.com/arvindsrinivas1/GlitterGallery

> If i wanted my parent to be original repo then I would have forked original repo

I agree. The parent should be the project that we `git clone` from, which in our case is the fork. It would be nice to show the user a dialog that says something like  "You're forking username/repo which was forked from originaluser/repo. Are you sure you don't want to fork the original repo instead?" - but it's not necessary.

**Comment by EmilyDirsh** [comment]
Fair enough. I think we do still want to maintain links between the forks and the original - similar to the Github network feature. We probably want to use https://github.com/stefankroes/ancestry or something similar to track the relationships.

**Comment by sonalkr132** [comment]
I guess I don't need to make any changes to my request? 

We can work on adding that, but I think we have more important things to achieve before adding details like that? I really want to see followers and notification feature working as soon as possible.

**Comment by EmilyDirsh** [comment]
I'd agree with that. I added #237 to track the network view as an enhancement. If you can just update the PR to fix the merge conflicts, we can merge it.

**Comment by sarupbanskota** [comment]
I may be wrong, but I _think_ GitHub used to behave the way Emily described. I distinctly remember trying it a year or so ago, when I was working out how to do this model. 

I'm not adding anything new to the conversation, but essentially we need a way to go back up to the parent. That's the reason why I thought about it this way to start with - back then on IRC, when I observed at how Fedora's design team works during the 'busy' days, it starts out with one person throwing out an idea, others picking it up and contributing to the original author's work. I've noticed that those who tend to pick up work from someone else who's already derived from the author (fork of fork), they have a hard time syncing because right now this stuff is manual and people don't see each other in time. Design work is hard to deal with in that sense - if you end up in 'merge conflicts', it's not easy to resolve them. That doesn't mean we shouldn't provide power to the forked projects however, as outlined in this (old) article: http://zbowling.github.io/blog/2011/11/25/github/

Fortunately, the solution to this is adopting the network view and alert, both of which @EmilyDirsh and @rohitpaulk have already pointed out. Another way to go about it is to let the network show up right below the project name.

@sonalkr132 you can revisit this later after you're done with the rest - because with your current fix, we don't have any problems per se :)

**Comment by EmilyDirsh** [comment]
Resolved in #240
